lib/PipelineC/Tracing/Runner.cpp

```diff
@@ -0,0 +1,429 @@
+      }
+    } else if constexpr (std::is_same_v<ReturnT, bool>) {
+      soft_assert(Ret == *llvm::yaml::parseBool(Return),
+                  "Return value differs from what happened in the trace");
+    }
+
+    if (Return.startswith(PointerPrefix)) {
+      if constexpr (std::is_pointer_v<ReturnT>) {
+        // Check that pointers are NULL or non-NULL
+        if (Return == NullPointer) {
+          revng_check(reinterpret_cast<uintptr_t>(Ret) == 0,
+                      "Function that returned a null pointer in the trace "
+                      "returned a non-null pointer");
+        } else {
+          soft_assert(reinterpret_cast<uintptr_t>(Ret) != 0,
+                      "Function that returned a non-null pointer in the trace "
+                      "returned a null pointer");
+        }
+
+        // Store the returned pointer
+        Context.storePointer(Return, Ret);
+      } else {
+        revng_abort();
+      }
+    }
+  }
+}
+
+template<ConstexprString Name, typename ReturnT, typename... Args>
+static CommandRunner makeRunner(std::function<ReturnT(Args...)> Function) {
+  return [Function](RunnerContext &Context,
+                    ArgumentsRef Arguments,
+                    ReturnRef Return) {
+    runnerImplementation<Name>(Function, Context, Arguments, Return);
+  };
+}
+
+template<ConstexprString Name, typename FunctionT>
+static void registerFunction(FunctionT Function) {
+  std::string StrName{ std::string_view(Name) };
+  if (!CommandHandler.contains(StrName)) {
+    CommandRunner Runner = makeRunner<Name>(std::function{ Function });
+    CommandHandler.insert({ StrName, Runner });
+  }
+}
+
+// rp_initialize causes an abort, this is done since there can only be one call
+// to it for the entire duration of the program and it is assumed that the
+// caller of Trace::run will have done so beforehand. We still allow traces to
+// have the first command be an rp_initialize for ease of use.
+static void handleRpInitialize(RunnerContext &Context,
+                               ArgumentsRef Arguments,
+                               ReturnRef Return) {
+  soft_assert(0, "rp_initialize invoked mid-trace");
+}
+
+// Ditto as above, but for the last instruction
+static void handleRpShutdown(RunnerContext &Context,
+                             ArgumentsRef Arguments,
+                             ReturnRef Return) {
+  soft_assert(0, "rp_shutdown invoked mid-trace");
+}
+
+static bool InitOk = []() {
+  // Add Overrides
+  CommandHandler.insert({ "rp_initialize", handleRpInitialize });
+  CommandHandler.insert({ "rp_shutdown", handleRpShutdown });
+  llvm::sys::AddSignalHandler(cleanTemporaryDirectories, nullptr);
+
+// Add autogenerated registerRunner calls to register the
+// other functions automatically
+#define FUNCTION(fname) registerFunction<#fname>(fname);
+#include "revng/PipelineC/Functions.inc"
+#undef FUNCTION
+
+  return true;
+}();
```

Comment:
Wrap `CommandHandler` in `class` and move this code into it constructor.
`registerFunction` should be a method of such class.

This `InitOk` pattern is always a code smell, don't do it please.

Reply:
Done

Reply:
I don't see it, please comment that something has been done/push only once you pushed it.